### PR TITLE
fix: [DHIS2-9405] Avoid leaking sensitive sql messages in case of sensitive exceptions like sqlgrammarexceptions etc (master)

### PR DESCRIPTION
**Pull request by ameenhere**

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java [hunk 1]
```diff
@@ -84,6 +85,11 @@
 @ControllerAdvice
 public class CrudControllerAdvice
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:86]: This Advice is only applied  to `MetadataSyncController` and `MetadataVersionController` so the "sensitive exception" logic would not be applied to Tracker import - which I guess was the original trigger for this PR.

Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:86]: I am confused now. ControllerAdvice is applied globally across all controllers unless explicitly limited correct? 
Why do you think this only applies to those 2 controllers? 
I have also tested them on `TrackedEntityInstanceController`  and the exception handlers are handling the exceptions for that as well.

Comment (luciano-fiandesio) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:86]: You are absolutely right: 
> By default, the methods in an @ControllerAdvice apply globally to all controllers

I got confused by the fact that we have a couple of controllers extending `CrudControllerAdvice`, which, could probably be corrected.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java [hunk 2]
```diff
@@ -84,6 +85,11 @@
...
 {
+    //Add sensitive exceptions into this array
+    private static final Class<?>[] SENSITIVE_EXCEPTIONS = { BadSqlGrammarException.class, org.hibernate.QueryException.class };
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:89]: I also wanted to handle the cases when these exceptions are nested inside other exceptions. For example `CompletionException` could have a nested exception  [`cause`] as `BadSqlGrammarException`. So wanted to eliminate adding extra exception handlers and then handling it separately for the nested ones. 
Does that justify my implementation? Feel free to recommend better ways to implement this to handle both cases.

[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:89]: So I double checked. Spring exception handlers does not catch "nested exceptions" if they are not nested by spring itself. So ideally the concrete exception that is actually thrown out the controller is what is handled. 
The PR currently handles this and can be easily extended to include more sensitive Exceptions as and when needed by adding the exception class into the list.

[the remark this answers is not in the document]
Comment (luciano-fiandesio) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java:89]: Got it, great that you checked. Then there is not much to discuss, your solution is the perfect approach for this case.

**Review by stian-sandvold — APPROVED** [review]

**Review by luciano-fiandesio — APPROVED** [review]

**Pull request merged**